zrml/parimutuel/src/lib.rs

```diff
@@ -0,0 +1,467 @@
+            ensure!(!winning_balance.is_zero(), Error::<T>::NoWinningShares);
+            debug_assert!(
+                outcome_total >= winning_balance,
+                "The outcome issuance should be at least as high as the individual balance of \
+                 this outcome!"
+            );
+
+            let pot_total = T::AssetManager::free_balance(market.base_asset, &pot_account);
+            let payoff_ratio_mul_base: BalanceOf<T> =
+                bdiv_floor(pot_total.saturated_into(), outcome_total.saturated_into())?
+                    .saturated_into();
+            let payoff: BalanceOf<T> = bmul_floor(
+                payoff_ratio_mul_base.saturated_into(),
+                winning_balance.saturated_into(),
+            )?
+            .saturated_into();
+
+            Self::check_values(
+                winning_balance,
+                pot_total,
+                outcome_total,
+                payoff_ratio_mul_base,
+                payoff,
+            )?;
+
+            let slashable_asset_balance = winning_balance;
+
+            T::AssetManager::slash(winning_asset, &who, slashable_asset_balance);
+
+            let remaining_bal = T::AssetManager::free_balance(market.base_asset, &pot_account);
+            let actual_payoff = payoff.min(remaining_bal);
+
+            T::AssetManager::transfer(market.base_asset, &pot_account, &who, actual_payoff)?;
+
+            Self::deposit_event(Event::RewardsClaimed {
+                market_id,
+                asset: winning_asset,
+                balance: slashable_asset_balance,
+                actual_payoff,
+                sender: who.clone(),
+            });
+
+            Ok(())
+        }
+
+        fn do_claim_refunds(who: T::AccountId, refund_asset: AssetOf<T>) -> DispatchResult {
+            let market_id = match refund_asset {
+                Asset::ParimutuelShare(market_id, _) => market_id,
+                _ => return Err(Error::<T>::NotParimutuelOutcome.into()),
+            };
+            let market = T::MarketCommons::market(&market_id)?;
+            ensure!(market.status == MarketStatus::Resolved, Error::<T>::MarketIsNotResolvedYet);
+            ensure!(market.scoring_rule == ScoringRule::Parimutuel, Error::<T>::InvalidScoringRule);
+            ensure!(
+                matches!(market.market_type, MarketType::Categorical(_)),
+                Error::<T>::NotCategorical
+            );
+            Self::market_assets_contains(&market, &refund_asset)?;
+            let winning_outcome = market.resolved_outcome.ok_or(Error::<T>::NoResolvedOutcome)?;
+            let pot_account = Self::pot_account(market_id);
+            let winning_asset = match winning_outcome {
+                OutcomeReport::Categorical(category_index) => {
+                    Asset::ParimutuelShare(market_id, category_index)
+                }
+                OutcomeReport::Scalar(_) => return Err(Error::<T>::NotCategorical.into()),
+            };
```

Comment:
Shared code with `do_claim_rewards`, could be moved into `get_winning_asset_type` for example.

Reply:
e1c9f5c180390f397ca3e9248f3f346c99d7f878